### PR TITLE
[WIP] Introduce kubeadm.yaml as config for kubeadm for v1.12 +

### PR DESCRIPTION
**Pull request by stephenmoloney**
***What does this commit/MR/PR do?***
    
 - Introduce kubeadm.yaml as config for kubeadm
- Allow conditional installation of config depending on kubeadm version
    
***Why is this commit/MR/PR needed?***
    
 - Allow configuration the kubeadm installation via yaml file

### CODE REVIEW

File: kubeadm/v1alpha3-config.yaml [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+certificatesDir: /etc/kubernetes/pki
+clusterName: kubernetes
+imageRepository: k8s.gcr.io
+
+
+---
+apiVersion: kubelet.config.k8s.io/v1beta1
+kind: KubeletConfiguration
+containerLogMaxFiles: 1
+containerLogMaxSize: 10Mi
```
Comment (stefanprodan) [inline kubeadm/v1alpha3-config.yaml]: I would make the `containerLogMaxSize` a TF variable with default to `100Mi`

Comment (stephenmoloney) [reply kubeadm/v1alpha3-config.yaml]: sure - will do.

**Comment by stefanprodan** [comment]
We should update the addons (dashboard and metrics server) to the latest version. This could be done in another PR but I don't think the current metrics server works with 1.12

File: kubeadm/v1beta1-config.yaml [hunk 2]
```diff
@@ -0,0 +1,52 @@
+---
+apiVersion: kubeadm.k8s.io/v1beta1
+kind: InitConfiguration
+localAPIEndpoint:
+  advertiseAddress: CONFIG_CLUSTER_PRIVATE_IP
+  bindPort: 6443
+nodeRegistration:
+  kubeletExtraArgs:
+    "feature-gates": "BlockVolume=true,CRIContainerLogRotation=true"
```
Comment (stephenmoloney) [inline kubeadm/v1beta1-config.yaml]: Having the featuregates as TF variables would be good too.
Just not sure how to make that variable

"BlockVolume=true,CRIContainerLogRotation=true"

featureGates:
  BlockVolume: true
  CRIContainerLogRotation: true

Some parsing logic would be needed to convert to yaml.

I suppose we could just add it later or have you any ideas?

Comment (stefanprodan) [reply kubeadm/v1beta1-config.yaml]: I would leave it for later.

**Comment by stephenmoloney** [comment]
> We should update the addons (dashboard and metrics server) to the latest version. This could be done in another PR but I don't think the current metrics server works with 1.12

I have another branch which has some drastic additions to this repo so I converted it to it's own repo as it was starting to diverge a bit too much to keep in a fork. I am thinking of opensourcing it but I'm happy to merge it in here either. Thing is  that it's quite opinionated - maybe chat about it on the flux chat channel some day if you like.

Here is a snippet off the `README.md`

```text
## Features

This will do the following:

  * reserves public IPs for each server or alternatively assigns designated reserved ip
  * provisions three bare-metal servers with Ubuntu 18.04
  * connects to the master server via SSH and installs Docker CE and kubeadm armhf apt packages
  * runs kubeadm init on the master server and configures kubectl
  * downloads the kubectl admin config file on your local machine and replaces the private IP with the public one
  * creates a Kubernetes secret with the Weave Net password
  * installs Weave Net with encrypted overlay
  * starts the nodes in parallel and installs Docker CE and kubeadm
  * joins the nodes in the cluster using the kubeadm token obtained from the master
  * if using a C2L server, the 250GB local ssd will be mounted
  * installs openisci (to enable openebs chart usage) on each node
  * installs helm tiller (v2)
  * labels one node as the edge node which will handle incoming requests through an ingress controller
  * Loads the `flux-ssh-deploy-key` secret into the master node via ssh
  * creates the flux private secret `flux-ssh-deploy-key`
  * Loads the `/tmp/sealed_secrets/tls.key` and `/tmp/sealed_secrets/tls.crt` secrets into the master node via ssh
  * creates the sealed secrets private secret sealed-secrets-key using the `tls.key` and `tls.crt`
  * A storage pool for openebs is created (a component required for openebs)
  * flux is deployed using the variable `flux_known_hosts` and the `flux-ssh-deploy-key` secret
  * once flux is deployed, all the charts and k8s manifests (including pre-existing sealed secrets)
    in the `k8s` directory are deployed via flux

## Acknowledgements

Please note that this project is forked from the excellent work by
[Stefan Prodan](https://github.com/stefanprodan) on
the [k8s-bare-metal project](https://github.com/stefanprodan/k8s-scw-baremetal)

The key differences are:

  - This project does not currently support scaleway arm servers
  - This project deploys [helm](https://github.com/helm/helm) by default
  - This project deploys [weaveworks flux](https://github.com/weaveworks/flux) by default
  - This project deploys [sealed-secrets](https://github.com/bitnami-labs/sealed-secrets) by default
  - This project deploys [open-ebs](https://github.com/openebs/openebs) by default

The aim of this project is to have a fully gitops-ready k8s cluster through
the use of a single `terraform apply` command with

  - persistent disk storage
  - secrets management
  - automated gitops with helm and flux
```


If you want me to make PR's for all or some of the above features - just let me know.

**Comment by stefanprodan** [comment]
I wander if we should drop support for ARM... Have you tested this PR on ARMv7 ?

**Comment by stephenmoloney** [comment]
> I wander if we should drop support for ARM... Have you tested this PR on ARMv7 ?

No, I only tested on C2S and C2L. None of the ARM servers.

**Comment by stephenmoloney** [comment]
> I wander if we should drop support for ARM...

There is no harm to leave it there. Maybe someone will take an interest later. Also, k8s itself might eventually allow heterogeneous arm/amd clusters.

Although, I do think that either:

- splitting if off into it's own branch
or 
- Having a different installation script

for the arm server is a good idea until arm has support for most things like helm.

**Comment by stefanprodan** [comment]
Kubernetes supports heterogeneous arm/amd nodes. The problem with Scaleway is that last time I've tried to create an ARM cluster the kernel wasn't working with kubelet anymore...

**Comment by stephenmoloney** [comment]
> Kubernetes supports heterogeneous arm/amd nodes. The problem with Scaleway is that last time I've tried to create an ARM cluster the kernel wasn't working with kubelet anymore...

I'll see if I get a chance to try the same later and revert back.

**Comment by stefanprodan** [comment]
Thanks a lot @stephenmoloney

**Comment by stephenmoloney** [comment]
@stefanprodan 
Tihs PR depends on all the other ones before it. It's cleaner to move it to a new PR.
Closing for #41

**Pull request closed without merging**